### PR TITLE
feat: added span name processor

### LINKED ISSUE #509: Specify span names
Currently, Vert.x core sets "operation" (HTTP Method for client/server span kinds) as a span name:

- https://github.com/eclipse-vertx/vertx-tracing/blob/master/vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/OpenTelemetryTracer.java#L75
- https://github.com/eclipse-vertx/vertx-tracing/blob/master/vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/OpenTelemetryTracer.java#L138

With an opened ticket for configuring the name https://github.com/eclipse-vertx/vertx-tracing/issues/59

which is a bit of non-descriptive. To be aligned with the other DIAL components, like chat, adapters, etc., which form the name as HTTP Method  + HTTP Path (e.g. "POST /path/to/resource"), the idea is to introduce the span processor which will set the name the same way.

### PR DESCRIPTION
**Pull request by alekseyvdovenko**
Currently, Vert.x core sets "operation" (HTTP Method for client/server span kinds) as a span name:

- https://github.com/eclipse-vertx/vertx-tracing/blob/master/vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/OpenTelemetryTracer.java#L75
- https://github.com/eclipse-vertx/vertx-tracing/blob/master/vertx-opentelemetry/src/main/java/io/vertx/tracing/opentelemetry/OpenTelemetryTracer.java#L138

With an opened ticket for configuring the name https://github.com/eclipse-vertx/vertx-tracing/issues/59

which is a bit of non-descriptive. To be aligned with the other DIAL components, like chat, adapters, etc., which form the name as HTTP Method  + HTTP Path (e.g. "POST /path/to/resource"), the idea is to introduce the span processor which will set the name the same way.

### CODE REVIEW

**Comment by alekseyvdovenko** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-510.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by alekseyvdovenko** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-510.nightly-test.deltixhub.io
>E2E tests status: failed

**Review by akurnosau — DISMISSED** [review]
I really like that "named group" approach

File: src/test/java/com/epam/aidial/core/util/RegexUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                        "/api/123"
+                ),
+                Arguments.of(
+                        "^/v1/(?<somethingElse>.+?)$",
+                        "/api/123",
+                        "/api/123"
+                )
+        );
+    }
+}
```
Comment (akurnosau) [inline src/test/java/com/epam/aidial/core/util/RegexUtilTest.java]: New line is missing here

Comment (alekseyvdovenko) [reply src/test/java/com/epam/aidial/core/util/RegexUtilTest.java]: added

**Comment by alekseyvdovenko** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-510.nightly-test.deltixhub.io
>E2E tests status: failed

File: src/main/java/com/epam/aidial/core/util/RegexUtil.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@UtilityClass
+public class RegexUtil {
+
+    public String replaceNamedGroups(Pattern pattern, String input, List<String> groups) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/util/RegexUtil.java:14]: can we improve the algorithm?

```
public String replaceNamedGroups(Pattern pattern, String input, List<String> groups) {
        if (groups == null || groups.isEmpty()) {
            return input;
        }
        List<RegexGroup> regexGroups = collectGroups(pattern, input, groups);
        if (regexGroups.isEmpty()) {
            return input;
        }
        regexGroups.sort(Comparator.comparingInt(a -> a.start));
        StringBuilder nameBuilder = new StringBuilder();
        int prev = 0;
        for (var rg : regexGroups) {
            nameBuilder.append(input, prev, rg.start);
            nameBuilder.append('{').append(rg.group).append('}');
            prev = rg.end;
        }
        nameBuilder.append(input, prev, input.length());
        return nameBuilder.toString();
    }
```

File: src/main/java/com/epam/aidial/core/util/RegexUtil.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        List<RegexGroup> regexGroups = new ArrayList<>();
+        Matcher matcher = pattern.matcher(input);
+        if (matcher.matches() && matcher.groupCount() > 0) {
+            for (String group : groups) {
+                try {
+                    int start = matcher.start(group);
+                    int end = matcher.end(group);
+                    regexGroups.add(new RegexGroup(group, start, end));
+                } catch (IllegalStateException | IllegalArgumentException ignored) {
+                    // Ignore
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/util/RegexUtil.java]: //Ignore group mismatch

File: src/main/java/com/epam/aidial/core/util/SpanUtil.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.List;
+import java.util.regex.Pattern;
+
+@UtilityClass
+public class SpanUtil {
+
+    private static final AttributeKey<String> ATTR_HTTP_METHOD = AttributeKey.stringKey(HttpTraceAttributeConstants.HTTP_METHOD);
+    private static final List<String> GROUPS = List.of("id", "bucket", "path");
+
+    public static void updateName(Pattern pathPattern, String path) {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/util/SpanUtil.java]: can we do re-factoring?

```
public static void updateName(Pattern pathPattern, String path, String httpMethod) {
        String spanName = RegexUtil.replaceNamedGroups(pathPattern, path, GROUPS);
        Span.current().updateName(httpMethod + " " + spanName);
    }
```

File: src/main/java/com/epam/aidial/core/util/RegexUtil.java (hunk 2 above)
Comment (alekseyvdovenko) [reply src/main/java/com/epam/aidial/core/util/RegexUtil.java:14]: thanks, your options looks better

File: src/main/java/com/epam/aidial/core/util/SpanUtil.java (hunk 4 above)
Comment (alekseyvdovenko) [reply src/main/java/com/epam/aidial/core/util/SpanUtil.java]: updated

File: src/main/java/com/epam/aidial/core/util/RegexUtil.java (hunk 2 above)
Comment (alekseyvdovenko) [reply src/main/java/com/epam/aidial/core/util/RegexUtil.java:14]: updated

File: src/main/java/com/epam/aidial/core/util/RegexUtil.java (hunk 3 above)
Comment (alekseyvdovenko) [reply src/main/java/com/epam/aidial/core/util/RegexUtil.java]: updated

**Comment by alekseyvdovenko** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-510.nightly-test.deltixhub.io
>E2E tests status: success

**Review by astsiapanay — APPROVED** [review]

**Pull request merged**